provisioner/data.js

```diff
@@ -86,6 +87,107 @@ var WorkerType = base.Entity.configure({
   context: ['provisionerId', 'keyPrefix'],
 });
 
+// We want to add a lastModified field
+WorkerType = WorkerType.configure({
+  version: 2,
+  properties: {
+    // These fields are documented in Version 1 of this Entity
+    workerType: base.Entity.types.String,
+    minCapacity: base.Entity.types.Number,
+    maxCapacity: base.Entity.types.Number,
+    scalingRatio: base.Entity.types.Number,
+    minPrice: base.Entity.types.Number,
+    maxPrice: base.Entity.types.Number,
+    canUseOndemand: base.Entity.types.JSON, // delete this
+    canUseSpot: base.Entity.types.JSON, // delete this
+    instanceTypes: base.Entity.types.JSON,
+    regions: base.Entity.types.JSON,
+    // Store the date of last modification for this entity
+    lastModified: base.Entity.types.Date,
+    // Global base UserData object for overwriting
+    userData: base.Entity.types.JSON,
+    // Global base LaunchSpecification object for overwriting
+    launchSpec: base.Entity.types.JSON,
+    // Global base secrets object for overwriting
+    secrets: base.Entity.types.JSON,
+    // Global base scope list for appending, these are the scopes that
+    // temporary TC credentials will be issued against
+    scopes: base.Entity.types.JSON,
+  },
+  migrate: function (item) {
+    console.log('Upgrading a workerType to version 2');
+    console.log(JSON.stringify(item, null, 2));
+    var oldUserData;
+    if (item.launchSpecification.UserData) {
+      oldUserData = JSON.parse(new Buffer(item.launchSpecification.UserData, 'base64').toString());
+    } else {
+      oldUserData = {};
+    }
+    var oldLaunchSpec = item.launchSpecification;
+    delete item.launchSpecification;
+    delete oldLaunchSpec.UserData;
+    item.launchSpec = lodash.clone(oldLaunchSpec);
+    item.secrets = {};
+    item.scopes = [];
+    item.userData = lodash.clone(oldUserData);
+
+    // Now, strip out the UserData and LaunchSpec stuff from existing entries
+    // and put them in the right structure then create empty scopes and secrets
+    item.regions.forEach(r => {
+      // Let's get the old launch spec and user data
+      var oldUserData;
+      if (r.overwrites && r.overwrites.UserData) {
+        oldUserData = JSON.parse(new Buffer(r.overwrites.UserData, 'base64').toString());
+        ['capacity', 'provisionerId', 'workerType'].forEach(key => {
+          if (oldUserData[key]) {
+            delete oldUserData[key];
+          }
+        });
+      } else {
+        oldUserData = {};
+      }
+      var oldOverwrites = r.overwrites;
+      delete r.overwrites;
+
+      // Now set them up
+      r.userData = oldUserData;
+      r.launchSpec = oldOverwrites;
+      r.secrets = {};
+      r.scopes = [];
+
+    });
+
+    item.instanceTypes.forEach(t => {
```

Comment:
Actually, since i'm now creating a new worker and using map, i can no longer handle instance types and regions the same as they have a slightly different set of properties.